### PR TITLE
losses: add likelihood loss and test, improve base loss

### PR DESCRIPTION
**Pull request by AldoGl**
## Proposed changes

I add a loss based on the computation of the likelihood function, I improve the modularity of the base loss class and I add a Jupyter notebook where the new loss is used to compute likelihood and posterior for the BH4 model.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] I have read the [CONTRIBUTING](../blob/master/CONTRIBUTING.md) doc
- [ ] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: black_it/loss_functions/likelihood.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+def kernel(sq_dist: NDArray[np.float64], h: float, D: int) -> NDArray[np.float64]:
+    """Compute a kernel density estimation using a Gaussian kernel."""
+    print(h)
+    k = np.exp(-(sq_dist / (2 * h**2))) / (h**D * (2 * np.pi) ** (D / 2.0))
+    return k
+
+
+class LikelihoodLoss(BaseLoss):
+    """Class for a likelihood loss computation using a Gaussian kernel function."""
```
Comment (marcofavoritobi) [inline black_it/loss_functions/likelihood.py:38]: OK for the first version using a Gaussian kernel function; but one might want to use a different density estimation approach?

Comment (marcofavoritobi) [inline black_it/loss_functions/likelihood.py:38]: we might want to add some references here if considered relevant (perhaps Grazzini et al 2017?)

Comment (AldoGl) [reply black_it/loss_functions/likelihood.py]: Yes, of course this can be useful, we can expand the class as necessary in the future

File: black_it/loss_functions/likelihood.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        self.h = h
+
+    @staticmethod
+    def _get_bandwidth_silverman(N: int, D: int) -> float:
+        """Return a reasonable bandwidth value computed using the Silverman's rule of thumb."""
+        h = ((N * (D + 2)) / 4) ** (-1 / (D + 4))
+        return h
+
+    @staticmethod
+    def _get_bandwidth_scott(N: int, D: int) -> float:
```
Comment (marcofavoritobi) [inline black_it/loss_functions/likelihood.py:66]: not used; maybe we can have a configuration in the loss (either scott or silverman)

File: tests/test_losses/test_likelihood.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+"""This module contains tests for the Likelihood loss."""
+import numpy as np
+
+from black_it.loss_functions.likelihood import LikelihoodLoss
+
+np.random.seed(3)
```
Comment (marcofavoritobi) [inline tests/test_losses/test_likelihood.py:21]: this is a bit dangerous as the line is executed at module level, but it is not guaranteed the seeds, across tests, are set in the same order because of changes in the order in how the test modules are imported. I suggest using a different approach; how is it done in other tests?

Comment (marcofavoritobi) [reply tests/test_losses/test_likelihood.py:21]: ok I see each test calls `np.random.seed` in the body of each test. I argue against this practice as `np.random.seed` does a global side-effect and might cause dependency among tests. This should be refactored in using a local random number generator (`np.default_rng(seed)`) whenever possible. 

For this PR, I suggest moving this call within the body of each test.

File: black_it/loss_functions/base.py [hunk 4]
```diff
@@ -62,6 +62,45 @@ def compute_loss(
...
+    def _filter_data(
+        num_coords: int,
+        filters: List[Optional[Callable]],
+        sim_data_ensemble: NDArray[np.float64],
+    ) -> NDArray[np.float64]:
+        """Filter the simulated time series."""
+        filtered_data = []
+
+        for i in range(num_coords):
+            filter_ = filters[i]
```
Comment (marcofavoritobi) [inline black_it/loss_functions/base.py:86]:
```suggestion
        for i, filter_ in enumerate(filters):
```

File: black_it/loss_functions/base.py [hunk 5]
```diff
@@ -62,6 +62,45 @@ def compute_loss(
...
+
+        loss = 0
+        for i in range(num_coords):
+            loss += self.compute_loss_1d(filtered_data[i], real_data[:, i]) * weights[i]
+
+        return loss
+
+    @staticmethod
+    def _filter_data(
+        num_coords: int,
```
Comment (marcofavoritobi) [inline black_it/loss_functions/base.py:78]: If suggestion below is accepted, `num_coords` is not needed here and above at line 68:

```suggestion
```

File: black_it/loss_functions/base.py [hunk 6]
```diff
@@ -62,6 +62,45 @@ def compute_loss(
         """
         num_coords = real_data.shape[1]
 
+        weights = self._check_coordinate_weights(num_coords)
+        filters = self._check_coordinate_filters(num_coords)
+
+        filtered_data = self._filter_data(num_coords, filters, sim_data_ensemble)
```
Comment (marcofavoritobi) [inline black_it/loss_functions/base.py:68]: See below

```suggestion
        filtered_data = self._filter_data(filters, sim_data_ensemble)
```

File: tests/test_losses/test_likelihood.py (hunk 3 above)
Comment (AldoGl) [reply tests/test_losses/test_likelihood.py:21]: I don't think we have a test with the same issue

File: black_it/loss_functions/base.py (hunk 4 above)
Comment (AldoGl) [reply black_it/loss_functions/base.py:86]: Sure, much better!

File: tests/test_losses/test_likelihood.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
+# Black-box ABM Calibration Kit (Black-it)
+# Copyright (C) 2021-2022 Banca d'Italia
```
Comment (marcofavoritobi) [inline tests/test_losses/test_likelihood.py:2]: we are in 2023, so we should change the year range of the licenses accordingly in all the copyright headers; but that's a task for another day...

Comment (AldoGl) [reply tests/test_losses/test_likelihood.py:2]: Right, I will change it for this module only in the meanwhile

File: tests/test_losses/test_likelihood.py (hunk 3 above)
Comment (AldoGl) [reply tests/test_losses/test_likelihood.py:21]: I can set the seeds within tests. In any case, the tests should work across different seeds

File: tests/test_losses/test_likelihood.py (hunk 7 above)
Comment (marcofavoritobi) [reply tests/test_losses/test_likelihood.py:2]: the copyright check will fail :( , however, you could add the year `2023` among the options in the regex of the script `scripts/check_copyright.py`.

In particular, at line 36, from:

```
# Copyright \(C\) 2021-2022 Banca d'Italia
```

To:

```
# Copyright \(C\) 2021-(2022|2023) Banca d'Italia
```

Comment (AldoGl) [reply tests/test_losses/test_likelihood.py:2]: Thanks 🙏🏼

**Review by marcofavoritobi — APPROVED** [review]
Great work :rocket:

**Comment by AldoGl** [comment]
Thank you @marcofavoritobi 🙏🏼
